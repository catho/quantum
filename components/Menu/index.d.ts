// Generated with scripts/create-component.js
import { FC } from 'react';

export interface MenuItemProps {
  id: string | number;
  content: React.ReactNode;
  handleClick: Function;
}

export interface MenuProps {
  open?: boolean;
  items?: MenuItemProps[];
  anchorEl?: Element | (() => Element) | null;
  anchorOrigin?: {
    vertical: 'top' | 'bottom' | 'center';
    horizontal: 'left' | 'center' | 'right';
  };
  transformOrigin?: {
    vertical: 'top' | 'bottom' | 'center';
    horizontal: 'left' | 'center' | 'right';
  };
  keepMounted?: boolean;
  handleClose?: (event: {}, reason: 'backdropClick' | 'escapeKeyDown') => void;
}

declare const Menu: FC<MenuProps>;
export default Menu;
