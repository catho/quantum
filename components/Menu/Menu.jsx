// Generated with scripts/create-component.js
import PropTypes from 'prop-types';
import MaterialMenu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import { createTheme } from '@mui/material/styles';
import { baseFontSize, colors } from '../shared/theme';

const materialThemeOverride = createTheme({
  typography: {
    fontFamily: `"Nunito Sans", sans-serif`,
    fontSize: baseFontSize,
    allVariants: {
      color: colors.neutral[700],
    },
  },
  palette: {
    text: {
      primary: colors.neutral[700],
    },
  },
});

const Menu = ({
  open = false,
  items = [],
  anchorEl = null,
  anchorOrigin = {
    vertical: 'bottom',
    horizontal: 'left',
  },
  transformOrigin = {
    vertical: 'top',
    horizontal: 'left',
  },
  keepMounted = false,
  onClose = () => {},
}) => (
  <MaterialMenu
    id="menu"
    data-testid="menu"
    theme={materialThemeOverride}
    aria-labelledby="button"
    anchorReference="anchorEl"
    anchorEl={anchorEl}
    open={open}
    onClose={onClose}
    onBlur={onClose}
    anchorOrigin={anchorOrigin}
    transformOrigin={transformOrigin}
    keepMounted={keepMounted}
  >
    {items.map((item) => {
      const onClickFunc = () => {
        const { onClick } = item;
        onClose();
        onClick();
      };
      return (
        <MenuItem
          key={item.id}
          onClick={onClickFunc}
          theme={materialThemeOverride}
          color="textPrimary"
        >
          {item.content}
        </MenuItem>
      );
    })}
  </MaterialMenu>
);

Menu.propTypes = {
  /** If true, the component is shown. */
  open: PropTypes.bool,
  /** Menu contents, has a content parameter, an id and a handleClick function. */
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
      content: PropTypes.node.isRequired,
      onClick: PropTypes.func,
    }),
  ),
  /**	An HTML element, or a function that returns one. It's used to set the position of the menu. */
  anchorEl: PropTypes.object,
  /** The point on the anchor where the popover's anchorEl will attach to.
   * - vertical: "top", "bottom", "center"
   * - horizontal: "left", "center", "right"
   * */
  anchorOrigin: PropTypes.object,
  /** The point on the popover which will attach to the anchor's origin.
   * - vertical: "top", "bottom", "center"
   * - horizontal: "left", "center", "right"
   * */
  transformOrigin: PropTypes.object,
  /** Always keep the children in the DOM. */
  keepMounted: PropTypes.bool,
  /** Callback fired when the component requests to be closed. */
  onClose: PropTypes.func,
};

export default Menu;
