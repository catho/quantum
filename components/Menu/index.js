// Generated with scripts/create-component.js
import Menu from './Menu';

export default Menu;
