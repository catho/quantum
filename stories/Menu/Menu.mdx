{/* Generated with scripts/create-componnent.js */}
import { Meta, Canvas, Story, ArgsTable } from '@storybook/addon-docs';
import  Menu  from '../../components/Menu';

import { Header } from '../shared';

import { 
  Default,
  WithIcon
} from './Menu.stories.jsx';

import EvaluationForm from '../../stories/shared/EvaluationForm'

<Meta title="Menu" component={Menu}  />

<Header title="Menu">
      A menu that displays a list of choices on a temporary surface. It appears when the user interacts with a button, or other control.
      It is based on the Menu Material UI component of the same name.
</Header>

- [Importing and usage](#importing)
- [API](#api)

## <a id="importing"></a>Importing and usage

```js
import { Menu } from '@catho/quantum';
```

```jsx
...
export default function BasicMenu() {
  const [anchorEl, setAnchorEl] = useState(null);
  const open = Boolean(anchorEl);
  const handleClick = (event) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <div>
      <Button
        aria-controls="presentation"
        aria-haspopup="true"
        aria-expanded={open}
        onClick={handleClick}
      >
        Menu
      </Button>
      <Menu
        open={open}
        anchorEl={anchorEl}
        onClose={handleClose}
        items={menuItems}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'left',
        }}
        {...args}
      />
    </div>
  );
}
```

<Canvas>
  <Story of={Default} />
</Canvas>

### With Icon

```jsx
...
export default function MenuWithIcon() {
  const [anchorEl, setAnchorEl] = useState(null);
  const open = Boolean(anchorEl);
  const handleClick = (event) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <div>
      <button
        type="button"
        aria-labelledby="Menu"
        aria-label="Menu"
        aria-controls="presentation"
        aria-haspopup="true"
        aria-expanded={open}
        onClick={handleClick}
      >
        <Icon name="menu" />{' '}
      </button>
      <Menu
        open={open}
        anchorEl={anchorEl}
        onClose={handleClose}
        items={menuItems}
        {...args}
      />
    </div>
  );
}
```

<Canvas withSource="none">
  <Story of={WithIcon}/>
</Canvas>

## <a id="api"></a>API

<ArgsTable of={Menu} />

<EvaluationForm componentName="Menu" />

    