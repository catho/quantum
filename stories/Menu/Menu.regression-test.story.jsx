// Generated with scripts/create-componnent.js

import { useState } from 'react';
import { Menu, Button } from '../../components';

export default {
  title: 'Menu',
  component: Menu,
};

const menuItems = [
  {
    id: 'test-candidate',
    content: 'Testar candidato',
    handleClick: () => {},
  },
  {
    id: 'print-cv',
    content: 'Imprimir currÃ­culo',
    handleClick: () => {},
  },
];

const Template = (args) => {
  const [anchorEl, setAnchorEl] = useState(null);
  const open = Boolean(anchorEl);
  const handleClick = (event) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <div>
      <Button
        id="button"
        aria-controls={open ? 'menu' : undefined}
        aria-haspopup="true"
        aria-expanded={open ? 'true' : undefined}
        onClick={handleClick}
      >
        Menu
      </Button>
      <Menu
        open={open}
        anchorEl={anchorEl}
        handleClose={handleClose}
        items={menuItems}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'left',
        }}
        {...args}
      />
    </div>
  );
};

export const Default = Template.bind({});
