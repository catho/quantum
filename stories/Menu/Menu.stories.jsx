// Generated with scripts/create-componnent.js
import { useState } from 'react';
import { Menu, Button, Icon } from '../../components';

export default {
  title: 'Menu',
  component: Menu,
};

const menuItems = [
  {
    id: 'test-candidate',
    content: 'Testar candidato',
    handleClick: () => {},
  },
  {
    id: 'print-cv',
    content: 'Imprimir currÃ­culo',
    handleClick: () => {},
  },
];

const Template = (args) => {
  const [anchorEl, setAnchorEl] = useState(null);
  const open = Boolean(anchorEl);
  const handleClick = (event) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <div>
      <Button
        id="button"
        aria-controls={open ? 'menu' : undefined}
        aria-haspopup="true"
        aria-expanded={open ? 'true' : undefined}
        onClick={handleClick}
      >
        Menu
      </Button>
      <Menu
        open={open}
        anchorEl={anchorEl}
        handleClose={handleClose}
        items={menuItems}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'left',
        }}
        {...args}
      />
    </div>
  );
};

const TemplateButtonIcon = (args) => {
  const [anchorEl, setAnchorEl] = useState(null);
  const open = Boolean(anchorEl);
  const handleClick = (event) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <div>
      <div
        aria-hidden="true"
        aria-label="Menu"
        role="button"
        id="button"
        aria-controls={open ? 'menu' : undefined}
        aria-haspopup="true"
        aria-expanded={open ? 'true' : undefined}
        onClick={handleClick}
      >
        <Icon name="menu" />{' '}
      </div>
      <Menu
        open={open}
        anchorEl={anchorEl}
        handleClose={handleClose}
        items={menuItems}
        {...args}
      />
    </div>
  );
};

export const Default = Template.bind({});
export const WithIcon = TemplateButtonIcon.bind({});
