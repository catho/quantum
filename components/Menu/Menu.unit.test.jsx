// Generated with scripts/create-component.js
import { render, fireEvent } from '@testing-library/react';
import Menu from './Menu';

describe('<Menu />', () => {
  const mockItems = [
    { id: 'item-1', content: 'Item 1', handleClick: jest.fn() },
    { id: 'item-2', content: 'Item 2', handleClick: jest.fn() },
  ];

  const mockAnchorEl = document.createElement('div');
  const mockHandleClose = jest.fn();

  const renderMenu = (props) => render(<Menu {...props} />);

  it('should match the snapshots', () => {
    expect(render(<Menu />).asFragment()).toMatchSnapshot();
  });
  it('renders with correct items', () => {
    const { getByText } = renderMenu({
      open: true,
      items: mockItems,
      anchorEl: mockAnchorEl,
      handleClose: mockHandleClose,
    });

    mockItems.forEach((item) => {
      const renderedItem = getByText(item.content);
      expect(renderedItem).toBeInTheDocument();
    });
  });

  it('calls handleClose when an item is clicked', () => {
    const { getByText } = renderMenu({
      open: true,
      items: mockItems,
      anchorEl: mockAnchorEl,
      handleClose: mockHandleClose,
    });

    mockItems.forEach((item) => {
      const renderedItem = getByText(item.content);
      fireEvent.click(renderedItem);
      expect(item.handleClick).toHaveBeenCalledTimes(1);
    });
  });

  it('renders with anchorEl when open is true', () => {
    const { getByTestId } = renderMenu({
      open: true,
      items: mockItems,
      anchorEl: mockAnchorEl,
      handleClose: mockHandleClose,
    });
    const menu = getByTestId('menu');
    expect(menu).toBeInTheDocument();
  });

  it('does not render with anchorEl when open is false', () => {
    const { queryByTestId } = renderMenu({
      open: false,
      items: mockItems,
      anchorEl: mockAnchorEl,
      handleClose: mockHandleClose,
    });
    const menu = queryByTestId('menu');
    expect(menu).toBeNull();
  });

  it('calls handleClose when onBlur event is triggered', () => {
    const { getByTestId } = renderMenu({
      open: true,
      items: mockItems,
      anchorEl: mockAnchorEl,
      handleClose: mockHandleClose,
    });
    const menu = getByTestId('menu');
    fireEvent.blur(menu);
    expect(mockHandleClose).toHaveBeenCalledTimes(1);
  });

  it('renders without items when items array is empty', () => {
    const { queryByText } = renderMenu({
      open: true,
      items: [],
      anchorEl: mockAnchorEl,
      handleClose: mockHandleClose,
    });
    mockItems.forEach((item) => {
      const renderedItem = queryByText(item.content);
      expect(renderedItem).toBeNull();
    });
  });
});
