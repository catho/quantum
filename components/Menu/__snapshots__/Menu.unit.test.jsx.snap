// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> should match the snapshots 1`] = `<DocumentFragment />`;
